File: objectiveai-sdk-rs/src/functions/inventions/http.rs
//! HTTP functions for function inventions.

use crate::{HttpClient, HttpError, McpHandler, Notifier};
use futures::Stream;

/// Creates a function invention (non-streaming).
pub async fn create_function_invention_unary(
    client: &HttpClient,
    mut params: super::request::FunctionInventionCreateParams,
) -> Result<super::response::unary::FunctionInvention, HttpError> {
    params.stream = None;
    client
        .send_unary(reqwest::Method::POST, "functions/inventions", Some(params))
        .await
}

/// Creates a streaming function invention. Returns
/// `(Stream<Chunk>, Notifier)`; see
/// [`crate::agent::completions::http::create_agent_completion_streaming`]
/// for the demux + handler semantics.
pub async fn create_function_invention_streaming<H: McpHandler>(
    client: &HttpClient,
    mut params: super::request::FunctionInventionCreateParams,
    handler: H,
) -> Result<
    (
        impl Stream<
            Item = Result<
                super::response::streaming::FunctionInventionChunk,
                HttpError,
            >,
        >
        + Send
        + Unpin
        + 'static
        + use<H>,
        Notifier,
    ),
    HttpError,
> {
    params.stream = Some(true);
    client
        .send_streaming_ws(
            reqwest::Method::POST,
            "functions/inventions",
            params,
            handler,
        )
        .await
}

File: objectiveai-sdk-py/objectiveai_sdk/functions/inventions/http.py
"""HTTP functions for function inventions."""

from __future__ import annotations

from typing import TYPE_CHECKING, Union

if TYPE_CHECKING:
    from objectiveai_sdk.client import ObjectiveAI
    from objectiveai_sdk.functions.inventions.request import FunctionInventionCreateParams
    from objectiveai_sdk.functions.inventions.response.streaming import FunctionInventionChunk
    from objectiveai_sdk.functions.inventions.response.unary import FunctionInvention
    from objectiveai_sdk.stream import Stream


async def create_function_invention(
    client: ObjectiveAI,
    params: FunctionInventionCreateParams,
) -> Union[FunctionInvention, Stream[FunctionInventionChunk]]:
    """Create a function invention.

    If ``params.stream`` is true, returns a streaming response.
    """
    if getattr(params, "stream", None):
        return await client.post_streaming("functions/inventions", params)
    return await client.post_unary("functions/inventions", params)
